test/tools/spec-runner/index.js

```diff
@@ -92,109 +93,123 @@ function gatherTestSuites(specPath) {
     let runOn = testSuite.runOn;
-    if (!testSuite.runOn) {
-      runOn = [{ minServerVersion: testSuite.minServerVersion }];
-      if (testSuite.maxServerVersion) {
-        runOn.push({ maxServerVersion: testSuite.maxServerVersion });
-      }
+    if (!testSuite.runOn && !Array.isArray(runOn)) {
+      throw new Error('no runOn requirement? it should be required');
     }
-    const environmentRequirementList = parseRunOn(runOn);
-
-    environmentRequirementList.forEach(requires => {
-      const suiteName = `${testSuite.name} - ${requires.topology.join()}`;
-      describe(suiteName, {
-        metadata: { requires },
-        test: function () {
-          beforeEach(() => prepareDatabaseForSuite(testSuite, testContext));
-          afterEach(() => testContext.cleanupAfterSuite());
-          testSuite.tests.forEach(spec => {
-            const maybeIt = shouldRunSpecTest(this.configuration, requires, spec, filter)
-              ? it
-              : it.skip;
-            maybeIt(spec.description, function () {
-              let testPromise = Promise.resolve();
-              if (spec.failPoint) {
-                testPromise = testPromise.then(() => testContext.enableFailPoint(spec.failPoint));
-              }
-
-              // run the actual test
-              testPromise = testPromise.then(() =>
-                runTestSuiteTest(this.configuration, spec, testContext)
-              );
-
-              if (spec.failPoint) {
-                testPromise = testPromise.then(() => testContext.disableFailPoint(spec.failPoint));
-              }
-              return testPromise.then(() => validateOutcome(spec, testContext));
-            });
-          });
-        }
-      });
-    });
-  });
-}
 
-function shouldRunSpecTest(configuration, requires, spec, filter) {
-  if (requires.authEnabled && process.env.AUTH !== 'auth') {
-    // TODO(NODE-3488): We do not have a way to determine if auth is enabled in our mocha metadata
-    // We need to do a admin.command({getCmdLineOpts: 1}) if it errors (code=13) auth is on
-    return false;
-  }
+    const beforeEachFilter = async function () {
+      let utilClient;
+      if (this.configuration.isLoadBalanced) {
+        // The util client can always point at the single mongos LB frontend.
+        utilClient = this.configuration.newClient(this.configuration.singleMongosLoadBalancerUri);
+      } else {
+        utilClient = this.configuration.newClient();
+      }
 
-  if (
-    requires.serverless &&
-    !shouldRunServerlessTest(requires.serverless, !!process.env.SERVERLESS)
-  ) {
-    return false;
-  }
+      await utilClient.connect();
 
-  if (
-    spec.operations.some(
-      op => op.name === 'waitForEvent' && op.arguments.event === 'PoolReadyEvent'
-    )
-  ) {
-    // TODO(NODE-2994): Connection storms work will add new events to connection pool
-    return false;
-  }
+      const allRequirements = runOn.map(legacyRunOnToRunOnRequirement);
+
+      let shouldRun = true;
```

Comment:
If you want to be unnecessarily fancy here (which I don't recommend, I just had fun writing this):

```typescript
const id = x => x;
let shoundRun = await Promise.all(
  allRequirements.map((req) => topologySatisfies(this.currentTest.ctx, requirement, utilClient))
).then(satisfied => satisfied.every(id))
```

Is it more readable?  Not sure.  Is it fun?  I think so 😄

Reply:
Clever approach! I think it would work the same way but we would get a nondeterministic "reason" depending on which runOn came back with false last. Something to consider 🤔